### PR TITLE
Fixes slow start of multiple worker processes (resolves #7136)

### LINKED ISSUE #7136: Worker processes do not start in linear time
The functions host adds worker processes more slowly than intended and as documented. Here are the details provided by @cgillum - 

```
According to this documentation, I should expect to wait 100 seconds before all 10 python worker processes are available.
https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_process_count

What I actually observed is that we take 10 seconds for the second python process to be created and 10 more seconds for each additional python worker process. Here is my Functions log output which illustrates what I mean: 

[2021-02-03T21:33:57.717] Worker process started and initialized.
[2021-02-03T21:34:07.667] Worker process started and initialized.
[2021-02-03T21:34:27.649] Worker process started and initialized.
[2021-02-03T21:34:57.652] Worker process started and initialized.
...
[2021-02-03T21:41:27.695] Worker process started and initialized.

It took 390 seconds before all 10 python worker processes came up.
```

### PR DESCRIPTION
**Pull request by alrod**
resolves #7136 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information
This is continuation of the PR:
https://github.com/Azure/azure-functions-host/pull/7138

### CODE REVIEW

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (alrod) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:231]: This can potentially improve cold start for non-placeholder scenarios as we are not waiting for the process to start.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (paulbatum) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: If you invert the order of these two operations - call the action first, then do task.delay, I think you can remove this startindex check.

**Review by paulbatum — CHANGES_REQUESTED** [review]
Can you add a new test case that fails without this fix? My previous PR had a unit test that confirmed the behavior change, though there was some feedback on the PR to use a different testing approach.

**Comment by alrod** [comment]
> Can you add a new test case that fails without this fix? My previous PR had a unit test that confirmed the behavior change, though there was some feedback on the PR to use a different testing approach.

I added some new code to existing tests as the tests already covers multiple language workers start up.

File: test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs
Comment (paulbatum) [inline test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs]: I would consider making the hardcoded values (10 seconds and 15 seconds) parameters, so that readers of the process start unit tests have an idea of what VerifyStartIntervals is doing without having to read through the implementation.

Comment (alrod) [reply test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs]: Fixed.

**Review by paulbatum — DISMISSED** [review]
Would be good to get one other approver before merging.

**Comment by pragnagopa** [comment]
> Would be good to get one other approver before merging.

Looking into this!

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Just make this constant with the proper comment/calculation.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: This task is unobserved. Please add a test inducing a failure in the code executed in the lambda and ensure that is properly handled. As-is, potential exceptions here will go unhandled in the background.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: No need to iterate. Just replace with `Any()`

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: nit: also update the comment as this is just scheduling a new `Task`.

**Review by fabiocav — CHANGES_REQUESTED** [review]

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: FYI - Failure to start worker process is handled here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs#L136-L147 - we have test coverage for this

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:231]: For the first process - we do want to block. Launching multiple  processes can be "fire and forget".
As good way to test this E2E - locally - Set up a queue with 100+ messages then start the functions runtime - check if executions succeed.

**Review by pragnagopa — COMMENTED** [review]
Added one comment. Rest looks good. Thanks!

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Can you clarify the comment above? Is that additional information to share that a class of errors is properly handled and surfaced, or are you suggesting we leave this as-is?

The code you've pointed to handles a number of known potential exceptions, but it does not make what we have here safe. There are a of calls up to that point that could introduce a failure that would end up being thrown in the background, and that's if we only think about the current state of the code. As some of those code paths evolve in the future, chances of exposing the fragility here just keep growing.

Any concerns with being defensive here?

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Sorry. I should have provided a complete response. The code I referred to above is an example of how to induce worker start up errors. Yes, if there is a way to log Task errors - we should

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (alrod) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:231]: I tested this with 100+ messages and all executions were succeed.
In old code we also set up `RpcFunctionInvocationDispatcher.State=FunctionInvocationDispatcherState.Initialized`; state in the non-blocking code:
https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs#L119

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:231]: Thanks for confirming!

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (alrod) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: can't make this constant as we set the value in workerConfig. I removed the assignment as it's calculated on InitializeAsync call.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (alrod) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Added the logging and a test.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Checking initialized state is handled here: https://github.com/Azure/azure-functions-host/blob/1927f5fd736931d2a6407b84d35927f2404eb8dc/src/WebJobs.Script/Description/Workers/WorkerFunctionInvoker.cs#L109-L123 

Safe to remove the check here.

Comment (alrod) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: The check is removed,

**Review by pragnagopa — APPROVED** [review]

**Review by fabiocav — APPROVED** [review]

**Pull request merged**
